vnext/local-cli/runWindows/utils/autolink.js

```diff
@@ -6,95 +6,341 @@
+    {
+        internal static void RegisterAutolinkedNativeModulePackages(IList<IReactPackageProvider> packageProviders)
+        {${csReactPacakgeProviders}
+        }
     }
+}
+`;
+
+      changesNecessary =
+        updateFile(csFile, '\uFEFF' + csContents, verbose, checkMode) ||
+        changesNecessary;
+    } else if (projectLang === 'cpp') {
+      let cppIncludes = '';
+      let cppPackageProviders = '';
+
+      for (const dependencyName in windowsDependencies) {
+        windowsDependencies[dependencyName].projects.forEach(project => {
+          if (project.directDependency) {
+            cppIncludes += `\n\n// Includes from ${dependencyName}`;
+            project.cppHeaders.forEach(header => {
+              cppIncludes += `\n#include <${header}>`;
+            });
 
-    //TODO: Update to the new RegisterAutolinkedNativeModulePackages method
-
-    let generatedIncludes =
-      '// AutolinkedNativeModules.g.h -- contents generated by "react-native run-windows"\\\r\n\\\r\n#pragma once\\\r\n\\\r\n';
-    const dependencies = config.dependencies;
-    let packageRegistrations =
-      '#define REGISTER_AUTOLINKED_NATIVE_MODULE_PACKAGES()';
-    for (const dependency in dependencies) {
-      const windowDependency = dependencies[dependency].platforms.windows;
-      const cppProjFile = windowDependency
-        ? windowDependency.cppProjFile
-        : null;
-      if (cppProjFile == null) {
-        console.log('No C++/WinRT project found for ' + dependency);
-        continue;
+            cppPackageProviders += `\n    // IReactPackageProviders from ${dependencyName}`;
+            project.cppPackageProviders.forEach(packageProvider => {
+              cppPackageProviders += `\n    packageProviders.Append(winrt::${packageProvider}());`;
+            });
+          }
+        });
       }
-      console.log(
-        `Adding include and package provider statement for ${dependency}`,
+
+      const cppFile = path.join(projectDir, 'AutolinkedNativeModules.g.cpp');
+
+      verboseMessage(
+        `Calculating ${chalk.bold(path.basename(cppFile))}...`,
+        verbose,
       );
-      const trimmedPackageName = windowDependency.packageName.trim();
-      const includeStatement = `\r\n#include <winrt/${trimmedPackageName}.h>`;
-      generatedIncludes += includeStatement;
-      const packageRegistration = `    \\\r\n    PackageProviders().Append(winrt::${trimmedPackageName}::ReactPackageProvider());`;
-      packageRegistrations += packageRegistration;
+
+      const cppContents = `// AutolinkedNativeModules.g.cpp contents generated by "react-native autolink-windows"
+#include "pch.h"
+#include "AutolinkedNativeModules.g.h"${cppIncludes}
+
+// clang-format off
+namespace winrt::Microsoft::ReactNative
+{
+
+void RegisterAutolinkedNativeModulePackages(winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::ReactNative::IReactPackageProvider> const& packageProviders)
+{${cppPackageProviders}
+}
+
+}
```

Comment:
same comment as C# of using a template file

Reply:
Done.